### PR TITLE
ci: conditionally skip python and frontend tests

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
This PR changes Python CI workflows to skip unit test execution for PRs that don't have changes that are relevant for those workflows. Here we're leveraging [`continue-on-error`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error) to check if files relevant to unit tests have changed, namely

Python changes:
- `tests/*`
- `superset/*`
- `requirements/*.txt`
- `setup.py`

Frontend changes:
- `superset_frontend/*`

This is done to ensure that skipped workflows return a successful status for required checks despite not executing the expensive test steps. If any of the aforementioned file have changed, a failure exit code is triggered, which subsequent steps listen in on; if the check step returned a failure, the tests run, otherwise they are skipped. 
To check changes, run `check_no_file_changes.sh` with the following arguments (the `GITHUB_REPO` and `PR_NUMBER` env variables need to be set):
- Only check python changes: `./check_no_file_changes.sh python`
- Only check frontend changes: `./check_no_file_changes.sh frontend`
- Check python and frontend changes: `./check_no_file_changes.sh python frontend` (order doesn't matter)

E2E tests have not been included in this PR, as some steps there depend in `failure()` which might need tweaking to run properly with this script.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Example of skipped frontend changes:
image
This shows the results for a PR with python changes:
```
Searching for changes in python files
Detected changes... Exiting with FAILURE code
bash-3.2$ PR_NUMBER=12960 GITHUB_REPO=apache/superset ./ci_check_no_file_changes.sh python
CHANGED FILES:
superset/migrations/versions/070c043f2fdb_add_granularity_to_charts_where_missing.py

Searching for changes in python files
Detected changes... Exiting with FAILURE code
bash-3.2$ PR_NUMBER=12960 GITHUB_REPO=apache/superset ./ci_check_no_file_changes.sh frontend
CHANGED FILES:
superset/migrations/versions/070c043f2fdb_add_granularity_to_charts_where_missing.py

Searching for changes in frontend files
No changes detected... Exiting with SUCCESS code
bash-3.2$ PR_NUMBER=12960 GITHUB_REPO=apache/superset ./ci_check_no_file_changes.sh python frontend
CHANGED FILES:
superset/migrations/versions/070c043f2fdb_add_granularity_to_charts_where_missing.py

```

When checking for both python and frontend changes on a PR with frontend changes, the following happens:
```
bash-3.2$ PR_NUMBER=12920 GITHUB_REPO=apache/superset ./ci_check_no_file_changes.sh python frontend
CHANGED FILES:
superset-frontend/spec/javascripts/sqllab/TableElement_spec.jsx
superset-frontend/src/SqlLab/components/SqlEditorLeftBar.jsx
superset-frontend/src/SqlLab/components/TableElement.jsx
superset-frontend/src/SqlLab/main.less
superset-frontend/src/common/components/Collapse/Collapse.stories.tsx
superset-frontend/src/common/components/Collapse/index.tsx
superset-frontend/src/common/components/common.stories.tsx
superset-frontend/src/common/components/index.tsx
superset-frontend/src/dashboard/components/FiltersBadge/DetailsPanel.tsx
superset-frontend/src/explore/components/ControlPanelsContainer.jsx
superset-frontend/src/explore/components/DataTablesPane.tsx
superset-frontend/src/explore/components/DatasourcePanel.tsx
superset-frontend/src/explore/components/controls/FixedOrMetricControl.jsx
superset-frontend/src/explore/main.less
superset-frontend/src/views/CRUD/welcome/Welcome.tsx

Searching for changes in python files
Searching for changes in frontend files
Detected changes... Exiting with FAILURE code
```

### TEST PLAN
Tested to correctly detect frontend and python changes locally, as well as skip tests when they are not needed to run

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by amitmiran137 — APPROVED** [review]
Wow this will reduce CI waiting time dramatically.
Thank you for that!

**Comment by junlincc** [comment]
@ktmud could you help review and stamp it? it's gonna be a big improvement 🙏

**Comment by ktmud** [comment]
You probably wouldn't want to ignore `tests/*.py` for E2E because it also contains some scripts to fill examples data.

I'm also wondering what's the difference between this solution and using Github Action's built-in [`on.<push|pull_request>.paths`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths) config?

We can probably get away with something like this:

```yaml
name: E2E

on:
  push:
    branches-ignore:
      - "dependabot/**/docs/**"
    paths-ignore:
      - "docs/**"
      - *.md
      - docker/**
      - helm/**
  pull_request:
    paths-ignore:
      - "docs/**"
      - *.md
      - docker/**
      - helm/**
```

```yaml
# Python MySQL unit tests
name: Python MySQL

on:
  push:
    branches-ignore:
      - "dependabot/npm_and_yarn/**"
    paths:
      - superset/**
      - tests/**
      - requirements/**
      - scripts/**
      - setup.py
  pull_request:
    paths:
      - superset/**
      - tests/**
      - requirements/**
      - scripts/**
      - setup.py
```

**Comment by nytai** [comment]
@ktmud using the github syntax causes the action to forever remain in a "pending" state which block PR merges for actions that are marked `required`. This approach causes the action to report back ✅ even if no tests were actually run.

**Review by nytai — APPROVED** [review]
LGTM. The python checks only took ~20s!

**Comment by ktmud** [comment]
> @ktmud using the github syntax causes the action to forever remain in a "pending" state which block PR merges for actions that are marked `required`. This approach causes the action to report back ✅ even if no tests were actually run.

Right. Totally forgot about this... We've talked about this before, lol.

**Comment by ktmud** [comment]
Anyway, I still don't think you can skip `test/*.py` for E2E tests.

File: .github/workflows/test-presto.yml [hunk 1]
```diff
@@ -42,11 +42,22 @@ jobs:
...
       - uses: actions/checkout@v2
+      - name: Check if python changes are present
+        id: check
+        env:
+          GITHUB_REPO: ${{ github.repository }}
+          PR_NUMBER: ${{ github.event.pull_request.number }}
+        continue-on-error: true
+        run: |
+          ./scripts/ci_check_python_changes.sh
+          ./scripts/ci_check_python_test_changes.sh
```
Comment (ktmud) [inline .github/workflows/test-presto.yml:53]: Can we maybe parameterize this script to avoid creating two largely identical files?

**Comment by nytai** [comment]
> Anyway, I still don't think you can skip `test/*.py` for E2E tests.

are we? I'm not seeing any changes to `superset-e2e.yml`

**Comment by ktmud** [comment]
> > Anyway, I still don't think you can skip `test/*.py` for E2E tests.
> 
> are we? I'm not seeing any changes to `superset-e2e.yml`

I was stated in the PR description as the reason for having two `.sh` files. Then maybe we don't need two files.

**Comment by ktmud** [comment]
BTW, we should probably also skip `superset-frontend.yml` for pure Python changes.

**Comment by nytai** [comment]
Ah, I didn't read the PR description fully. I agree, since e2e is supposed to be out final catch-all check it should probably always run. Also there is code in `tests/*.py` that can affect e2e test runs

**Comment by villebro** [comment]
Thanks for the reviews and good comments @nytai and @ktmud , I'll update the PR accordingly

File: .github/workflows/test-presto.yml (hunk 1 above)
Comment (villebro) [reply .github/workflows/test-presto.yml:53]: > > Anyway, I still don't think you can skip `test/*.py` for E2E tests.
> 
> are we? I'm not seeing any changes to `superset-e2e.yml`

I wanted to keep the scope of this PR as small as possible, hence didn't het add the E2E changes. Once this is merged I'll add similar logic to frontend and E2E tests.

**Comment by villebro** [comment]
> BTW, we should probably also skip `superset-frontend.yml` for pure Python changes.

Yes, I'll do this in a follow-up PR.

File: .github/workflows/test-presto.yml (hunk 1 above)
Comment (ktmud) [reply .github/workflows/test-presto.yml:53]: Another benefit of parameterizing this is to allow more advanced skipping in the future---for example, we may split/skip E2E tests based on product areas.

File: scripts/ci_check_python_changes.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+FILES=$(curl -s -X GET -G $URL | jq -r '.[] | .filename')
+
+cat<<EOF
+CHANGED FILES:
+$FILES
+
+EOF
+
+if [[ "${FILES}" =~ (superset\/.+\.py|setup\.py|requirements\/.+\.txt) ]]; then
+  echo "Detected changes in python files... Exiting with FAILURE code"
```
Comment (dpgaspar) [inline scripts/ci_check_python_changes.sh:30]: nit: seems logically odd that a script named `check_python_changes` returns "False" if there were changes, assuming it's easier to detect `if: steps.check.outcome == 'failure'` maybe rename the script to a negative

**Review by dpgaspar — APPROVED** [review]
LGTM, just a non blocking comment, agree that E2E should run also on python changes

File: scripts/ci_check_python_changes.sh (hunk 2 above)
Comment (villebro) [reply scripts/ci_check_python_changes.sh:30]: That's a good point also. I'll rename the files to be in line with what's being asserted.

**Comment by ktmud** [comment]
The `ci_check_no_file_changes.sh` file seems to be missing!

File: scripts/ci_check_no_file_changes.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+#
+
+# Usage instructions:
+#
+# to check for python changes, run with CHECKS=python
+# To check for frontend changes, run with CHECKS=frontend
+# To check for python and frontend changes, run with CHECKS="python frontend"
+
+URL="https://api.github.com/repos/${GITHUB_REPO}/pulls/${PR_NUMBER}/files"
+FILES=$(curl -s -X GET -G $URL | jq -r '.[] | .filename')
```
Comment (dpgaspar) [inline scripts/ci_check_no_file_changes.sh:27]: nit: `SC2086: Double quote to prevent globbing and word splitting.`

**Review by dpgaspar — APPROVED** [review]
This is great @villebro, one step forward to a faster CI

**Pull request merged**